### PR TITLE
group0_state_machine: use correct comparison for timeuuids in `merger`

### PR DESCRIPTION
**Pull request by kbr-scylla**
In d2a4079bbeef21d051d07e317d871ad54efd1896, `merger` was modified so that when we merge a command, `last_group0_state_id` is taken to be the maximum of the merged command's state_id and the current `last_group0_state_id`. This is necessary for achieving the same behavior as if the commands were applied individually instead of being merged -- where we take the maximum state ID from `group0_history` table which was applied until now (because the table is sorted using the state IDs and we take the greatest row).

However, a subtle bug was introduced -- the `std::max` function uses the `utils::UUID` standard comparison operator which is unfortunately not the same as timeuuid comparison that Scylla performs when sorting the `group0_history` table. So in rare cases it could return the *smaller* of the two timeuuids w.r.t. the correct timeuuid ordering. This would then lead to commands being applied which should have been turned to no-ops due to the `prev_state_id` check -- and then, for example, permanent schema desync or worse.

Fix it by using the correct comparison method.

Fixes: #14600

### CODE REVIEW

**Comment by kbr-scylla** [comment]
I'll let @margdoc write a unit test to catch this as part of https://github.com/scylladb/scylladb/issues/14568 as an award for giving me a nice puzzle which I spent too much time trying to crack today.

**Review by tgrabiec — APPROVED** [review]

File: service/raft/group0_state_machine.cc [hunk 1]
```diff
@@ -121,7 +121,14 @@ future<> group0_state_machine::apply(std::vector<raft::command_cref> command) {
         void add(group0_command&& cmd, size_t added_size) {
             slogger.trace("add to merging set new_state_id: {}", cmd.new_state_id);
             auto m = convert_history_mutation(std::move(cmd.history_append), sm._sp.data_dictionary());
-            last_group0_state_id = std::max(last_group0_state_id, cmd.new_state_id);
+            // Set `last_group0_state_id` to the maximum of the current value and `cmd.new_state_id`,
+            // but make sure we compare them the same way timeuuids are compared in clustering keys
+            // (i.e. in the same order that the history table is sorted).
+            auto cmp = timeuuid_type->compare(
+                timeuuid_type->decompose(last_group0_state_id), timeuuid_type->decompose(cmd.new_state_id));
```
Comment (kostja) [inline service/raft/group0_state_machine.cc:128]: I understand the urgency of the fix, but adding two decompose and then a virtual call for comparing two integers is too much. Could you please implement UUID::timeuuid_tricompare following the impplementation of timeuuid_tri_compare function?

**Comment by kbr-scylla** [comment]
v2: avoid allocations/virtual calls when comparing the timeuuids.

A new timeuuid comparison function was introduced that operates directly on `const UUID&`s.

**Review by kostja — APPROVED** [review]

**Comment by kbr-scylla** [comment]
> You could amend this comment and point to the timeuuid_tri_compare function. This will make it easier to find and use a proper comparator.

Good point. I pushed a new commit.

**Review by margdoc — APPROVED** [review]

**Comment by avikivity** [comment]
We should have a timeuuid C++ type and use it where convenient.

In fact we have timeuuid_native_type, but it's too high up in the food chain. We need a timeuuid in UUID.hh, and then timeuuid_native_type can alias it.

**Comment by bhalevy** [comment]
> We should have a timeuuid C++ type and use it where convenient.
> 
> In fact we have timeuuid_native_type, but it's too high up in the food chain. We need a timeuuid in UUID.hh, and then timeuuid_native_type can alias it.

Yes, I think we can do that if we make `operator<=>` virtual and override it in the timeuuid derived class.

**Comment by kbr-scylla** [comment]
> CI state FAILURE - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2399/

known flaky test (I'll work on it today), restarted build

**Comment by kbr-scylla** [comment]
@tgrabiec please merge

**Pull request merged**